File: odd-or-even-sum/odd-or-even-sum-1.hs
-- This solution times out on HackerRank

doQuery :: [Int] -> (Int, Int) -> String
doQuery numbers (from, to) =
  if odd $ foldl (\a n -> (abs a) + n) x xs
    then "Odd"
    else "Even"
  where
    x:xs = take (to - from + 1) $ drop (from - 1) numbers

solve :: ([Int], [(Int, Int)]) -> [String]
solve (numbers, queries) = (doQuery numbers) <$> queries

parseQuery :: String -> (Int, Int)
parseQuery queryStr = (read a, read b)
  where
    [a, b] = words queryStr

parse :: String -> ([Int], [(Int, Int)])
parse input = (num, queries)
  where
    numStr:queriesStr = tail $ lines input
    num = read <$> words numStr :: [Int]
    queries = parseQuery <$> queriesStr :: [(Int, Int)]

main = interact $ unlines . solve . parse

File: odd-or-even-sum/odd-or-even-sum-1.rs
// This solutions passes all tests on HackerRank, even if the algorithm is NxQ
fn main() {
    use std::io::{self, BufRead};

    let stdin = io::stdin();

    let mut lines = stdin.lock().lines().map(Result::unwrap).skip(1);

    let numbers: Vec<i32> = lines
        .next()
        .unwrap()
        .split(' ')
        .map(|x| x.parse().unwrap())
        .collect();

    let queries: Vec<(usize, usize)> = lines
        .map(|query_str| {
            let q: Vec<usize> = query_str.split(' ').map(|x| x.parse().unwrap()).collect();
            (q[0], q[1])
        })
        .collect();

    for (from, to) in queries {
        let slice = &numbers[from-1..to];
        let x : i32 = slice.iter().fold(0, |acc, n| acc.abs() + n);
        println!("{}", if x & 1 == 1 { "Odd" } else { "Even" });
    }
}
